### PR TITLE
Implemented FadeProgress class

### PR DESCRIPTION
**Pull request by oxixes**
Any comments or improvements are highly appreciated.

- - -
This change is [Reviewable](https://reviewable.io/reviews/zeldaret/botw/110)

### CODE REVIEW

File: src/Game/UI/uiFadeProgress.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    void init();
+    void reset();
+    void updateFadeScreen();
+    void setProgress(float value);
+    float getGaugeValue();
+    void setIsActive();
+
+private:
+    bool mActive = false;
+    f32 mProgress = 0.0;
```
Comment (leoetlino) [inline src/Game/UI/uiFadeProgress.h:22]: Could we use float instead of f32 (for consistency)?

File: src/Game/UI/uiFadeProgress.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    mGaugeValue = 0.0;
+    mSteps = 1.0;
+}
+
+void FadeProgress::reset() {
+    init();
+}
+
+void FadeProgress::updateFadeScreen() {
+    if (mActive) {
```
Comment (leoetlino) [inline src/Game/UI/uiFadeProgress.cpp:20]:
```
if (!mActive)
    return;
```
(rationale: https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code)

File: src/Game/UI/uiFadeProgress.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            mSteps = 1.1;
+        if (difference > 0.2f)
+            mSteps = 1.2;
+        if (progress > 0.94f)
+            mSteps = 10.0;
+
+        if (difference <= 0.0001f && difference >= -0.0001f)
+            mSteps = 1.0;
+
+        f32 newProgress = mGaugeValue + (mProgressPerStep * mSteps);
```
Comment (leoetlino) [inline src/Game/UI/uiFadeProgress.cpp:34]: `new_progress` (Nintendo uses snake_case for local variables. yes, this is despite member variables being mWrittenLikeThis)

**Review by leoetlino — COMMENTED** [review]
mostly LGTM, I only have several minor comments

File: src/Game/UI/uiFadeProgress.h (hunk 1 above)
Comment (oxixes) [reply src/Game/UI/uiFadeProgress.h:22]: As I saw other parts of the code use the defined types in `seadTypes.h`, I thought it was better to use them, but sure.

Comment (leoetlino) [reply src/Game/UI/uiFadeProgress.h]: Yeah both float and f32 are used in the codebase as we're pretty bad at being consistent across the entire codebase :sweat_smile:

float is used a lot more though and shows up 5215 times (whereas f32 only shows up 1170 times).

Ultimately I think it doesn't really matter and I think Nintendo probably uses both in their codebase (considering some class names use F32 and others use Float) but it'd be nice to stay consistent within a given source file at least

File: src/Game/UI/uiFadeProgress.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    SEAD_SINGLETON_DISPOSER(FadeProgress)
+    FadeProgress() = default;
+    ~FadeProgress() = default;
+
+public:
+    void init();
+    void reset();
+    void updateFadeScreen();
+    void setProgress(float value);
+    float getGaugeValue();
```
Comment (leoetlino) [inline src/Game/UI/uiFadeProgress.h:17]: Sorry, missed this earlier, but could you make this a const member function?
```suggestion
    float getGaugeValue() const;
```

File: data/uking_functions.csv [hunk 5]
```diff
@@ -49384,13 +49384,13 @@ Address,Quality,Size,Name
...
-0x00000071007cbae0,U,000180,FadeProgress::updateFadeScreen
-0x00000071007cbb94,U,000020,FadeProgress::setProgress
-0x00000071007cbba8,U,000008,FadeProgress::getGaugeValue
-0x00000071007cbbb0,U,000012,FadeProgress::setIsActive
+0x00000071007cba28,O,000144,_ZN5uking2ui12FadeProgress14createInstanceEPN4sead4HeapE
+0x00000071007cbab8,O,000020,_ZN5uking2ui12FadeProgress4initEv
+0x00000071007cbacc,O,000020,_ZN5uking2ui12FadeProgress5resetEv
+0x00000071007cbae0,O,000180,_ZN5uking2ui12FadeProgress16updateFadeScreenEv
+0x00000071007cbb94,O,000020,_ZN5uking2ui12FadeProgress11setProgressEf
+0x00000071007cbba8,O,000008,_ZN5uking2ui12FadeProgress13getGaugeValueEv
```
Comment (leoetlino) [inline data/uking_functions.csv:49392]:
```suggestion
0x00000071007cbba8,O,000008,_ZNK5uking2ui12FadeProgress13getGaugeValueEv
```
(the constness of a member functions affects its symbol mangling)

**Review by leoetlino — APPROVED** [review]

**Pull request merged**